source/tutorials/declarative-and-reproducible-developer-environments.rst

```diff
@@ -1,61 +1,151 @@
+.. _declarative-reproducible-envs:
+
 Declarative and reproducible developer environments
 ===================================================
 
-Nix can create reproducible environments given a declarative
-configuration called a Nix expression.
+In :ref:`ad-hoc-envs` tutorial we took a dive into providing shell
+environments for when we need a quick'n'dirty way of getting hold
+of some tools.
+
+In this tutorial we'll take a look how to create :term:`reproducible`
+shell environments given a declarative configuration file called a Nix expression.
+
+
+When are declarative shell environments useful?
+-----------------------------------------------
+
+This is the quickest approach to getting started with Nix:
+
+- single command to invoke it via ``nix-shell``
+- works across different operating systems (Linux / MacOS)
+- share the exact same environment with all developers
 
-Reproducible means you can share
-the configuration with others and guarantee that they are using the same
-software as you.
+Developer environments allow you to:
 
-To get started, make a new folder and create a file called ``shell.nix``
-with the following contents:
+- provide CLI tools, such as ``psql``, ``jq``, ``tmux``, etc
+- provide developer libraries, such as ``zlib``, ``openssl``, etc
+- set shell environment variables
+- execute bash during environment activation
+
+
+An example
+----------
+
+In top-level of your project create ``shell.nix`` with the following contents:
 
 .. code:: nix
 
-   { pkgs ? import <nixpkgs> {} }:
+   { pkgs ? import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/3590f02e7d5760e52072c1a729ee2250b5560746.tar.gz") {} }:
```

Comment:
Assuming that the reader hasn't been introduced to nixpkgs pinning at this point, that line probably looks very obscure. How is that URL generated? `{}` being used in that line? Why is the `pkgs` attribute optional?

Maybe say something in the lines of: "Don't worry about the content, for now, all of this will be explained in later chapters.".

Expanding the line might also help:

```suggestion
   let
      nixpkgs = fetchTarball "https://github.com/NixOS/nixpkgs/archive/3590f02e7d5760e52072c1a729ee2250b5560746.tar.gz";
   in
   { pkgs ? import nixpkgs {} }:
```

Reply:
Chronologically, pinning nixpkgs is before this tutorial (although it's still recommended to be read here).

I don't like `import nixpkgs` vs `import <nixpkgs>` as it's very confusing for a beginner.

Maybe it's best to mention nixpkgs tutorial here as a way to understand what's going on?

Reply:
Maybe a reference is enough: "If you don't understand this line, look at the [previous chapter](link)"